File: Android/app/src/main/java/com/auralis/crisisconnect/screens/Tools/RecentDisastersViewModel.kt
package com.auralis.crisisconnect.screens.Tools

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.auralis.crisisconnect.screens.Tools.data.CountryLocator
import com.auralis.crisisconnect.screens.Tools.data.DisasterErrorType
import com.auralis.crisisconnect.screens.Tools.data.DisasterEvent
import com.auralis.crisisconnect.screens.Tools.data.DisasterFeed
import com.auralis.crisisconnect.screens.Tools.data.DisasterRegion
import com.auralis.crisisconnect.screens.Tools.data.DisasterRepository
import com.auralis.crisisconnect.screens.Tools.data.haversineDistanceKm
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

private const val NEARBY_RADIUS_KM = 500.0

data class RecentDisastersUiState(
    val isLoading: Boolean = false,      // initial load / tab switch (full-screen skeleton)
    val isRefreshing: Boolean = false,   // pull-to-refresh (keeps current list visible)
    val error: DisasterErrorType? = null,// only set when the list is empty due to a failure
    val isStale: Boolean = false,        // showing cached data after a failed refresh
    val lastUpdatedMillis: Long = 0L,
    val selectedTab: String = "Global",  // "Global" (GDACS) | "Local" (AFAD)
    val selectedCategory: String = "All",
    val selectedAlertLevel: String = "All",
    val searchQuery: String = "",
    val detailEvent: DisasterEvent? = null,
    val hasLocation: Boolean = false,
    val nearbyEvents: List<DisasterEvent> = emptyList(),
    val otherEvents: List<DisasterEvent> = emptyList(),
    val hasAnyEvents: Boolean = false,   // any events loaded (pre-filter) — distinguishes "no data" from "no match"
    val primarySource: String = DisasterRepository.SOURCE_GDACS,
    val hasRegionalTab: Boolean = false, // the "your country" tab exists once a country is detected
    val regionalTabLabel: String = ""    // detected country name shown on that tab
)

class RecentDisastersViewModel(application: Application) : AndroidViewModel(application) {

    private val repository = DisasterRepository(application)
    private val countryLocator = CountryLocator(application)

    private val _uiState = MutableStateFlow(RecentDisastersUiState())
    val uiState: StateFlow<RecentDisastersUiState> = _uiState.asStateFlow()

    /** Full result set for the active feed, before filtering. */
    private var allEvents: List<DisasterEvent> = emptyList()
    private var userLat: Double? = null
    private var userLon: Double? = null

    /** The user's detected country; null until/unless detection succeeds. */
    private var region: DisasterRegion? = null

    init {
        viewModelScope.launch {
            // Detect the user's country first so the default tab can be theirs.
            val detected = countryLocator.detect()
            if (detected != null && detected.countryName.isNotBlank()) {
                region = detected
                _uiState.update {
                    it.copy(
                        hasRegionalTab = true,
                        regionalTabLabel = detected.countryName,
                        selectedTab = "Local" // default to the user's own country
                    )
                }
            }
            loadData(forceRefresh = false)
        }
    }

    fun loadData(forceRefresh: Boolean = false) {
        val isGlobal = _uiState.value.selectedTab == "Global"
        val feed = if (isGlobal) DisasterFeed.GLOBAL else DisasterFeed.LOCAL
        val activeRegion = if (isGlobal) null else region

        viewModelScope.launch {
            _uiState.update {
                if (forceRefresh) it.copy(isRefreshing = true)
                else it.copy(isLoading = true, error = null)
            }

            val result = repository.load(feed, activeRegion, forceRefresh)
            allEvents = result.events

            _uiState.update {
                it.copy(
                    isLoading = false,
                    isRefreshing = false,
                    isStale = result.isStale,
                    error = if (result.events.isEmpty()) (result.error ?: DisasterErrorType.NO_DATA) else null,
                    lastUpdatedMillis = result.lastUpdatedMillis,
                    hasAnyEvents = result.events.isNotEmpty(),
                    primarySource = result.primarySource
                )
            }
            recompute()
        }
    }

    fun changeTab(tabName: String) {
        if (_uiState.value.selectedTab == tabName) return
        allEvents = emptyList()
        _uiState.update {
            it.copy(
                selectedTab = tabName,
                searchQuery = "",
                selectedCategory = "All",
                selectedAlertLevel = "All",
                nearbyEvents = emptyList(),
                otherEvents = emptyList(),
                hasAnyEvents = false
            )
        }
        loadData(forceRefresh = false)
    }

    fun selectCategory(category: String) {
        _uiState.update { it.copy(selectedCategory = category) }
        recompute()
    }

    fun selectAlertLevel(level: String) {
        _uiState.update { it.copy(selectedAlertLevel = level) }
        recompute()
    }

    fun updateSearchQuery(query: String) {
        _uiState.update { it.copy(searchQuery = query) }
        recompute()
    }

    fun showDetail(event: DisasterEvent?) {
        _uiState.update { it.copy(detailEvent = event) }
    }

    /** Called by the screen whenever a fresh device location is obtained. */
    fun updateLocation(lat: Double, lon: Double) {
        userLat = lat
        userLon = lon
        recompute()
    }

    /** Rebuilds the filtered + nearby/other partition from [allEvents]. */
    private fun recompute() {
        val state = _uiState.value
        val filtered = allEvents.filter { event ->
            val categoryMatches = state.selectedCategory == "All" ||
                event.eventType.equals(state.selectedCategory, ignoreCase = true)
            val levelMatches = state.selectedAlertLevel == "All" ||
                event.alertLevel.equals(state.selectedAlertLevel, ignoreCase = true)
            val searchMatches = state.searchQuery.isBlank() ||
                event.country.contains(state.searchQuery, ignoreCase = true) ||
                event.title.contains(state.searchQuery, ignoreCase = true)
            categoryMatches && levelMatches && searchMatches
        }

        val lat = userLat
        val lon = userLon
        if (lat == null || lon == null) {
            _uiState.update {
                it.copy(hasLocation = false, nearbyEvents = emptyList(), otherEvents = filtered)
            }
            return
        }

        // Compute each distance once, then partition.
        val withDistance = filtered
            .filter { it.latitude != 0.0 && it.longitude != 0.0 }
            .map { it to haversineDistanceKm(lat, lon, it.latitude, it.longitude) }

        val nearby = withDistance
            .filter { it.second <= NEARBY_RADIUS_KM }
            .sortedBy { it.second }
            .map { it.first }

        val nearbySet = HashSet(nearby)
        val others = filtered.filter { it !in nearbySet }

        _uiState.update {
            it.copy(hasLocation = true, nearbyEvents = nearby, otherEvents = others)
        }
    }
}

File: iOS/Crisis Connect/Features/RecentDisasters/RecentDisastersViewModel.swift
//
//  RecentDisastersViewModel.swift
//  Crisis Connect
//
//  Ported from the Android RecentDisastersViewModel. Detects the user's country
//  for the default "local" tab, loads merged disaster feeds, and partitions
//  results into a nearby/other split based on device location.
//

import Foundation
import Combine
import CoreLocation
import WidgetKit

private let nearbyRadiusKm = 500.0

@MainActor
final class RecentDisastersViewModel: ObservableObject {

    // Loading / error state
    @Published var isLoading = false        // initial load / tab switch (full-screen skeleton)
    @Published var isRefreshing = false     // pull-to-refresh (keeps current list visible)
    @Published var error: DisasterErrorType? // only set when the list is empty due to a failure
    @Published var isStale = false          // showing cached data after a failed refresh
    @Published var lastUpdatedMillis: Int64 = 0

    // Filters / selection
    @Published var selectedTab = "Global"   // "Global" | "Local"
    @Published var selectedCategory = "All"
    @Published var selectedAlertLevel = "All"
    @Published var searchQuery = ""
    @Published var detailEvent: DisasterEvent?

    // Derived content
    @Published var hasLocation = false
    @Published var nearbyEvents: [DisasterEvent] = []
    @Published var otherEvents: [DisasterEvent] = []
    @Published var hasAnyEvents = false     // any events loaded (pre-filter)
    @Published var primarySource = DisasterSource.gdacs

    // Regional ("your country") tab
    @Published var hasRegionalTab = false
    @Published var regionalTabLabel = ""

    private let repository = DisasterRepository()
    private let countryLocator = CountryLocator()

    private var allEvents: [DisasterEvent] = []
    private var userLat: Double?
    private var userLon: Double?
    private var region: DisasterRegion?
    private var didStart = false

    var hasEvents: Bool { !nearbyEvents.isEmpty || !otherEvents.isEmpty }

    /// Called once when the screen appears. Detects the user's country (so the
    /// default tab can be theirs) and performs the first load.
    func start(initialCoordinate: CLLocationCoordinate2D?) async {
        guard !didStart else { return }
        didStart = true

        let detected = await countryLocator.detect(coordinate: initialCoordinate)
        if let detected, !detected.countryName.isEmpty {
            region = detected
            hasRegionalTab = true
            regionalTabLabel = detected.countryName
            selectedTab = "Local" // default to the user's own country
            // Hand the detected region to the home-screen widget (App Group).
            WidgetSharedState.saveRegion(detected)
        }
        await loadData(forceRefresh: false)
    }

    func loadData(forceRefresh: Bool) async {
        let isGlobal = selectedTab == "Global"
        let feed: DisasterFeed = isGlobal ? .global : .local
        let activeRegion = isGlobal ? nil : region

        if forceRefresh {
            isRefreshing = true
        } else {
            isLoading = true
            error = nil
        }

        let result = await repository.load(feed: feed, region: activeRegion, forceRefresh: forceRefresh)
        allEvents = result.events

        isLoading = false
        isRefreshing = false
        isStale = result.isStale
        error = result.events.isEmpty ? (result.error ?? .noData) : nil
        lastUpdatedMillis = result.lastUpdatedMillis
        hasAnyEvents = !result.events.isEmpty
        primarySource = result.primarySource
        recompute()

        if !result.events.isEmpty {
            WidgetCenter.shared.reloadTimelines(ofKind: WidgetSharedState.disastersWidgetKind)
        }
    }

    func changeTab(_ tabName: String) {
        guard selectedTab != tabName else { return }
        allEvents = []
        selectedTab = tabName
        searchQuery = ""
        selectedCategory = "All"
        selectedAlertLevel = "All"
        nearbyEvents = []
        otherEvents = []
        hasAnyEvents = false
        Task { await loadData(forceRefresh: false) }
    }

    func selectCategory(_ category: String) {
        selectedCategory = category
        recompute()
    }

    func selectAlertLevel(_ level: String) {
        selectedAlertLevel = level
        recompute()
    }

    func updateSearchQuery(_ query: String) {
        searchQuery = query
        recompute()
    }

    func showDetail(_ event: DisasterEvent?) {
        detailEvent = event
    }

    /// Called whenever a fresh device location is obtained.
    func updateLocation(latitude: Double, longitude: Double) {
        userLat = latitude
        userLon = longitude
        recompute()
    }

    /// Rebuilds the filtered + nearby/other partition from `allEvents`.
    private func recompute() {
        let filtered = allEvents.filter { event in
            let categoryMatches = selectedCategory == "All"
                || event.eventType.uppercased() == selectedCategory.uppercased()
            let levelMatches = selectedAlertLevel == "All"
                || event.alertLevel.uppercased() == selectedAlertLevel.uppercased()
            let query = searchQuery.trimmingCharacters(in: .whitespacesAndNewlines)
            let searchMatches = query.isEmpty
                || event.country.localizedCaseInsensitiveContains(query)
                || event.title.localizedCaseInsensitiveContains(query)
            return categoryMatches && levelMatches && searchMatches
        }

        guard let lat = userLat, let lon = userLon else {
            hasLocation = false
            nearbyEvents = []
            otherEvents = filtered
            return
        }

        let withDistance = filtered
            .filter { $0.hasCoordinates }
            .map { ($0, DisasterGeo.haversineKm(lat, lon, $0.latitude, $0.longitude)) }

        let nearby = withDistance
            .filter { $0.1 <= nearbyRadiusKm }
            .sorted { $0.1 < $1.1 }
            .map { $0.0 }

        let nearbyIds = Set(nearby.map { $0.id })
        let others = filtered.filter { !nearbyIds.contains($0.id) }

        hasLocation = true
        nearbyEvents = nearby
        otherEvents = others
    }
}

// ─── Location provider ─────────────────────────────────────────────────────────

/// Lightweight CoreLocation wrapper for the Recent Disasters screen. Provides a
/// best last-known location (for the nearby split) and requests "when in use"
/// permission on demand. Mirrors the screen-level location handling on Android.
@MainActor
final class RecentDisastersLocationProvider: NSObject, ObservableObject, CLLocationManagerDelegate {
    @Published var currentLocation: CLLocation?
    @Published var authorizationStatus: CLAuthorizationStatus

    private let manager = CLLocationManager()

    override init() {
        authorizationStatus = manager.authorizationStatus
        super.init()
        manager.delegate = self
        manager.desiredAccuracy = kCLLocationAccuracyHundredMeters
    }

    var isAuthorized: Bool {
        authorizationStatus == .authorizedAlways || authorizationStatus == .authorizedWhenInUse
    }

    /// Returns the best immediately-available location, if any.
    var bestLastKnownCoordinate: CLLocationCoordinate2D? {
        (currentLocation ?? (isAuthorized ? manager.location : nil))?.coordinate
    }

    /// Requests permission if needed and refreshes the location.
    func requestLocation() {
        switch authorizationStatus {
        case .notDetermined:
            manager.requestWhenInUseAuthorization()
        case .authorizedAlways, .authorizedWhenInUse:
            refresh()
        default:
            break
        }
    }

    func refresh() {
        guard isAuthorized else { return }
        if let location = manager.location {
            currentLocation = location
        }
        manager.requestLocation()
    }

    nonisolated func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        let status = manager.authorizationStatus
        Task { @MainActor in
            self.authorizationStatus = status
            if self.isAuthorized { self.refresh() }
        }
    }

    nonisolated func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }
        Task { @MainActor in
            self.currentLocation = location
        }
    }

    nonisolated func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        // Best-effort: keep whatever location we already have.
    }
}
